### DISCUSSION TITLE
Passing a String to image widget as filename

### DISCUSSION THREAD

**Discussion by Ultimariolator** [missionpinball/missionpinball#65] [Q&A] [answered]
I would like to make a player variable dependant image show. image widget says that it expects a string as the image name. But it will not accept a string from variable.
Here is the example extracted where I would like to show the image named "picture". It works when I type directly the picture:

player_vars:
  name:
    initial_value: 'picture'
    value_type: str

slides:
  event:
    widgets:
      - type: image
        image: (name)

So is this how it is or am I perhaps missing something? Is there any way to get the image defined other than directly specifying?

**Comment by atummons** [comment, marked as answer]
https://docs.missionpinball.org/en/stable/displays/widgets/reusable_widgets.html#dynamically-choosing-a-widget-based-on-variables

Look at this section. You are missing the syntax (current_player.name).

You have to tell it what name is, as that could exist multiple places. This tells it to look at the player variables for the current player to return name. Then it can pass through.

**Reply by Ultimariolator** [reply]
I had tried the suggested synatx but tried again to be sure. So the following did not work either:

...
- type: image
        image: (current_player.Name)

I also tried without parenthesis. But doesn't work. MPF and screen start but screen crashes when the widget is called.
The example you linked will probably work but is built by using variable widgets where each widget is defined separately and would result in at least 3 rows per alternative image where a direct reference from variable would allow me to get away with only 1 widget for all. But is the multiple widgets the only way?

**Reply by atummons** [reply]
Try (current_player.name). Lowercase name, since it is case sensitive.  If that doesn't work, can you post the error?

**Reply by Ultimariolator** [reply]
Thanks for the suggestion. That also causes it to crash. I had thought if it was due to asset pools so I used a direct image. But that also ends the same way.

I won't clutter this with the whole logs but these lines from the end pretty much sum it up I believe:

mpf.core.events.EventHandlerException: Exception while processing RegisteredHandler(callback=<bound method McConfigPlayer.play_from_trigger of McConfigPlayer.slides>, priority=1, kwargs={}, key=UUID('79a44084-acde-4820-a942-bd2f3593c194'), condition=None, blocking_facility=None) for event slides_play. Cannot add Image widget. Image '(current_player.name)' is not a valid image name.

Also tried with (player1|name):

mpf.core.events.EventHandlerException: Exception while processing RegisteredHandler(callback=<bound method McConfigPlayer.play_from_trigger of McConfigPlayer.slides>, priority=1, kwargs={}, key=UUID('0de158e4-1dda-41b9-8d3a-ffcf5bbe4983'), condition=None, blocking_facility=None) for event slides_play. Cannot add Image widget. Image '(player1|name)' is not a valid image name.

And tried with '(name)' just because it works with the text widget.

mpf.core.events.EventHandlerException: Exception while processing RegisteredHandler(callback=<bound method McConfigPlayer.play_from_trigger of McConfigPlayer.slides>, priority=1, kwargs={}, key=UUID('fb8a813a-8e0f-4a4f-8709-6576aafd9488'), condition=None, blocking_facility=None) for event slides_play. Cannot add Image widget. Image '(name)' is not a valid image name.

**Reply by atummons** [reply]
So I guess that's your answer. Since you're trying to add the image straight, it needs to be the name of the image itself.

You could follow the example and create a widget for each example. Create widget A tied to image a.png, B tied to image b.png, etc.

Then instead of image in your slide reference the widget, which will then call an image by its file name. What you have now is looking for a file called (name), which doesn't exist.

**Reply by Ultimariolator** [reply]
Any alternative suggestions? Could I use tokens or something similar for the image widget. Thing is I am aiming to have over 100 different images, so naturally 100 different dynamically chosen widgets still results in 300 config rows compared to somehow being able to directly pass the variable to the image name.
But I guess I will brute force it if there is no other way.
In any case, thanks for the help.

**Reply by Stefku** [reply]
Hi @Ultimariolator, I have a quite similar problem with the image name. I took another approach though: see #59 . There I tried the same as you but with tokens.
Interestingly: I am able to parameterize the text of an text widget, but _not the_ image of an image widget.

I got kind of similar error message like you: ` Cannot add Image widget. Image '(image_name)' is not a valid image name.`. For me it looks as the placeholder inside of parenthesis is ignored for the image widget. I also tried to debug the MC but I don't know how (see #58)

**Reply by Ultimariolator** [reply]
Yep, so it appears to be set in stone that an image widget can't be parametrised.
I also had no problem to use a string variable with a text widget.

**Reply by atummons** [reply]
Yeah, I got that far in code review the other day. I will look at adding placeholders to images. I'll see what I can do and report back.
